File: backend/src/main/java/ar/edu/utn/frc/previsar/services/EstructuraService.java
package ar.edu.utn.frc.previsar.services;

import ar.edu.utn.frc.previsar.dtos.DocumentoRequeridoDto;
import ar.edu.utn.frc.previsar.dtos.EstructuraExpedienteDto;
import ar.edu.utn.frc.previsar.dtos.SeccionDto;
import ar.edu.utn.frc.previsar.dtos.request.DocumentoRequeridoRequestDto;
import ar.edu.utn.frc.previsar.dtos.request.SeccionRequestDto;

/**
 * Administración de la estructura documental (secciones y documentos requeridos)
 * de un (provincia, tipo_tarea).
 *
 * La lectura es pública para cualquier usuario autenticado; las operaciones de
 * escritura solo las puede hacer un revisor, y únicamente sobre la estructura de
 * la provincia que ocupa (ver RolRevisor).
 */
public interface EstructuraService {

    /** Estructura vigente de un tipo de tarea en una provincia. */
    EstructuraExpedienteDto obtenerEstructura(Long tipoTareaId, Long provinciaId);

    /**
     * Estructura para el armado de un expediente propio: incluye los documentos
     * vigentes MÁS los que el expediente ya tiene cargados aunque hayan sido
     * retirados de la estructura (marcados con {@code desactivado=true}). Valida que
     * el expediente pertenezca al usuario (404 ante ajenos).
     */
    EstructuraExpedienteDto obtenerEstructuraParaExpediente(Long expedienteId);

    /**
     * Estructura vigente de un tipo de tarea en la provincia del revisor autenticado.
     * Para la pantalla de configuración: el revisor solo ve/edita su propia provincia.
     */
    EstructuraExpedienteDto obtenerEstructuraDeMiProvincia(Long tipoTareaId);

    SeccionDto crearSeccion(SeccionRequestDto request);

    SeccionDto actualizarSeccion(Long seccionId, SeccionRequestDto request);

    /** Baja lógica de la sección y de sus documentos. */
    void eliminarSeccion(Long seccionId);

    DocumentoRequeridoDto crearDocumento(Long seccionId, DocumentoRequeridoRequestDto request);

    DocumentoRequeridoDto actualizarDocumento(Long documentoId, DocumentoRequeridoRequestDto request);

    void eliminarDocumento(Long documentoId);

    /**
     * Copia la estructura de {@code provinciaOrigenId} hacia la provincia del
     * revisor autenticado. Devuelve la cantidad de secciones clonadas.
     */
    int clonarEstructura(Long tipoTareaId, Long provinciaOrigenId);
}

File: frontend/src/app/core/services/estructura.service.ts
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs/internal/Observable';
import {
  DocumentoRequerido,
  DocumentoRequeridoRequest,
  EstructuraExpediente,
} from '../models/estructura.model';
import { HttpClient, HttpParams } from '@angular/common/http';
import { API } from '../constants/api.constants';

@Injectable({
  providedIn: 'root',
})
export class EstructuraService {
  private readonly http = inject(HttpClient);

  getEstructura(tipoTareaId: number, provinciaId: number): Observable<EstructuraExpediente> {
    const params = new HttpParams()
      .set('tipoTareaId', tipoTareaId)
      .set('provinciaId', provinciaId);
    return this.http.get<EstructuraExpediente>(API.ESTRUCTURA, { params });
  }

  /**
   * Estructura para el armado de un expediente propio: incluye las ranuras retiradas
   * de la estructura vigente que este expediente ya tiene cargadas (desactivado = true).
   */
  getEstructuraExpediente(expedienteId: number): Observable<EstructuraExpediente> {
    return this.http.get<EstructuraExpediente>(API.ESTRUCTURA_EXPEDIENTE(expedienteId));
  }

  /** Estructura de la provincia del revisor autenticado (pantalla de configuración). */
  getEstructuraMia(tipoTareaId: number): Observable<EstructuraExpediente> {
    const params = new HttpParams().set('tipoTareaId', tipoTareaId);
    return this.http.get<EstructuraExpediente>(API.ESTRUCTURA_MIA, { params });
  }

  crearDocumento(
    seccionId: number,
    body: DocumentoRequeridoRequest,
  ): Observable<DocumentoRequerido> {
    return this.http.post<DocumentoRequerido>(API.ESTRUCTURA_SECCION_DOCUMENTOS(seccionId), body);
  }

  actualizarDocumento(
    documentoId: number,
    body: DocumentoRequeridoRequest,
  ): Observable<DocumentoRequerido> {
    return this.http.put<DocumentoRequerido>(API.ESTRUCTURA_DOCUMENTO(documentoId), body);
  }

  /** Baja lógica (soft delete): el backend nunca borra el archivo físico. */
  eliminarDocumento(documentoId: number): Observable<void> {
    return this.http.delete<void>(API.ESTRUCTURA_DOCUMENTO(documentoId));
  }
}
